### PR TITLE
Fixed a regression with reporting unused parameters in potential predicates

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/58493

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot test it

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -11360,18 +11360,23 @@ export function createNameResolver({
                         location = root.parent;
                     }
                     break;
-                case SyntaxKind.Parameter:
+                case SyntaxKind.Parameter: {
+                    const param = location as ParameterDeclaration;
```
Comment (sandersn) [inline src/compiler/utilities.ts:11364]: style question: is there a reason for all the casts in this code? Maybe it's better to drop this alias and use `location as ParameterDeclaration)` everywhere.

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -11360,18 +11360,23 @@ export function createNameResolver({
                         location = root.parent;
                     }
                     break;
-                case SyntaxKind.Parameter:
+                case SyntaxKind.Parameter: {
+                    const param = location as ParameterDeclaration;
+                    if (lastLocation === param.name) {
+                        lastSelfReferenceLocation = param;
```
Comment (sandersn) [inline src/compiler/utilities.ts:11366]: I'd almost prefer a new parameter and entry in `isSelfReferenceLocation` like

```ts
    function isSelfReferenceLocation(node: Node, lastLocation: Node): node is SelfReferenceLocation {
        switch (node.kind) {
            case SyntaxKind.Parameter: return lastLocation === (node as ParameterDeclaration).name
            // ...
```

File: src/compiler/utilities.ts [hunk 3]
```diff
@@ -11360,18 +11360,23 @@ export function createNameResolver({
                         location = root.parent;
                     }
                     break;
-                case SyntaxKind.Parameter:
+                case SyntaxKind.Parameter: {
+                    const param = location as ParameterDeclaration;
+                    if (lastLocation === param.name) {
```
Comment (sandersn) [inline src/compiler/utilities.ts:11365]: what does this do when param.name is a BindingPattern? Do we have tests to make sure that they still correctly get marked unused (and used when used)?

**Review by sandersn — CHANGES_REQUESTED** [review]
@danvk guessed that the source of the problem was a speculative `getTypeOfSymbol`. But this PR fixes resolution for all parameters. If it applies to all parameters, why wasn't this observed before Dan's change?

1. Need tests of used/unused binding patterns. These probably exist already, but please double-check.
2. If this implementation is correct, I think it would be easier to read in `isSelfReferenceLocation`, as long as that doesn't make performance worse (from adding a second parameter).

**Comment by Andarist** [comment]
> My (fuzzier) guess is that we haven't tried to resolve a parameter name from itself before and never observed that it mistakenly caused the parameter to be marked as used.

This is correct.

> But I'm not sure what in the new code is causing that.

It's `checkIfExpressionRefinesParameter` through the `getFlowTypeOfReference`. It leads to `narrowTypeByCallExpression` and thus to `isMatchingReference` and `getResolvedSymbol` that calls `resolveName`. It's this newly introduced "fake" control flow that changed this. By fake, I mean it's not created by the binder.

File: src/compiler/utilities.ts (hunk 1 above)
Comment (Andarist) [reply src/compiler/utilities.ts:11364]: No specific reason, I just prefer to have a single cast than multiple ones that are the same for the same variable. I can change this if you prefer the other style.

File: src/compiler/utilities.ts (hunk 2 above)
Comment (Andarist) [reply src/compiler/utilities.ts]: I just pushed out the suggested change

File: src/compiler/utilities.ts (hunk 3 above)
Comment (Andarist) [reply src/compiler/utilities.ts]: yeah, there are tests for that already and the changes to `checkIfExpressionRefinesAnyParameter` that introduced this bug don't affect binding patterns

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -11417,7 +11417,7 @@ export function createNameResolver({
                         location = root.parent;
                     }
                     break;
-                case SyntaxKind.Parameter:
+                case SyntaxKind.Parameter: {
```
Comment (jakebailey) [inline src/compiler/utilities.ts:11420]: Nit: leftover formatting change?

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -11429,6 +11429,7 @@ export function createNameResolver({
                         }
                     }
                     break;
+                }
```
Comment (Andarist) [inline src/compiler/utilities.ts:11432]:
```suggestion
```

File: src/compiler/utilities.ts (hunk 4 above)
Comment (Andarist) [reply src/compiler/utilities.ts:11420]:
```suggestion
                case SyntaxKind.Parameter:
```

**Comment by jakebailey** [comment]
@typescript-bot test it
@typescript-bot pack this

**Review by jakebailey — APPROVED** [review]

**Review by sandersn — APPROVED** [review]

**Comment by jakebailey** [comment]
@typescript-bot cherry-pick this to release-5.5

**Pull request merged**